src/lighteval/tasks/registry.py

```diff
@@ -75,63 +76,44 @@ def __init__(self, cache_dir: str):
-    ) -> LightevalTask:
+        # Private attributes, not expected to be mutated after initialization
+        self._cache_dir = cache_dir
+        self._custom_tasks = custom_tasks
+
+    def get_task_class(self, task_name: str):
         """
         Get the task class based on the task name.
 
         Args:
             task_name (str): Name of the task.
-            custom_tasks_registry (Optional[dict[str, LightevalTask]]): A dictionary containing custom tasks.
-
         Returns:
             LightevalTask: Task class.
 
         Raises:
             ValueError: If the task is not found in the task registry or custom task registry.
         """
-        if task_name in self.TASK_REGISTRY:
-            if custom_tasks_registry is not None and task_name in custom_tasks_registry:
-                hlog_warn(
-                    f"One of the tasks you requested ({task_name}) exists both in the default and custom tasks. Selecting the default task."
-                )
-            return self.TASK_REGISTRY[task_name]
-        if custom_tasks_registry is not None and task_name in custom_tasks_registry:
-            return custom_tasks_registry[task_name]
-        hlog_warn(f"{task_name} not found in provided tasks")
-        hlog_warn(pformat(self.TASK_REGISTRY))
-        raise ValueError(
-            f"Cannot find tasks {task_name} in task list or in custom task registry ({custom_tasks_registry})"
-        )
-
-    def get_task_dict(
-        self, task_name_list: List[str], custom_tasks: Optional[Union[str, Path, ModuleType]] = None
-    ) -> Dict[str, LightevalTask]:
-        """
-        Get a dictionary of tasks based on the task name list.
-
-        Args:
-            task_name_list (List[str]): A list of task names.
-            custom_tasks (Optional[Union[str, ModuleType]]): Path to the custom tasks file or name of a module to import containing custom tasks or the module itself
-            extended_tasks (Optional[str]): The path to the extended tasks group of submodules
+        task_class = self.task_registry.get(task_name)
+        if task_class is None:
+            hlog_warn(f"{task_name} not found in provided tasks")
+            hlog_warn(pformat(self.task_registry))
+            raise ValueError(f"Cannot find tasks {task_name} in task list or in custom task registry)")
 
-        Returns:
-            Dict[str, LightevalTask]: A dictionary containing the tasks.
+        return task_class
 
-        Notes:
-            - If custom_tasks is provided, it will import the custom tasks module and create a custom tasks registry.
-            - Each task in the task_name_list will be instantiated with the corresponding task class.
+    @property
+    @lru_cache
+    def task_registry(self):
         """
+        Returns dict mapping all available suite|task into LightevalTask
+        """
+
         # Import custom tasks provided by the user
```

Comment:
Very nice use of `lru_cache` - we'll need to update the code base with it for few-shot samples pools too for example